hoomd/hpmc/ExternalFieldHarmonic.h

```diff
@@ -0,0 +1,468 @@
+            energy += calcE(i, position, orientation);
+            }
+
+#ifdef ENABLE_MPI
+        if (this->m_sysdef->isDomainDecomposed())
+            {
+            MPI_Allreduce(MPI_IN_PLACE,
+                          &energy,
+                          1,
+                          MPI_HOOMD_SCALAR,
+                          MPI_SUM,
+                          m_exec_conf->getMPICommunicator());
+            }
+#endif
+
+        return energy;
+        } // end getEnergy(uin64_t)
+
+    //! Calculate the change in energy from moving a single particle with tag = index
+    double energydiff(const unsigned int& index,
+                      const vec3<Scalar>& position_old,
+                      const Shape& shape_old,
+                      const vec3<Scalar>& position_new,
+                      const Shape& shape_new)
+        {
+        double old_U = calcE(index, position_old, shape_old),
+               new_U = calcE(index, position_new, shape_new);
+        return new_U - old_U;
+        }
+
+    protected:
+    //! Calculate the energy associated with the deviation of a single particle from its reference
+    //! position
+    Scalar calcE_trans(const unsigned int& index, const vec3<Scalar>& position)
+        {
+        ArrayHandle<unsigned int> h_tags(m_pdata->getTags(),
+                                         access_location::host,
+                                         access_mode::read);
+        int3 dummy = make_int3(0, 0, 0);
+        vec3<Scalar> origin(m_pdata->getOrigin());
+        const BoxDim& box = this->m_pdata->getGlobalBox();
+        vec3<Scalar> r0 = m_reference_positions[h_tags.data[index]];
+        Scalar3 t = vec_to_scalar3(position - origin);
+        box.wrap(t, dummy);
+        vec3<Scalar> shifted_pos(t);
+        vec3<Scalar> dr = vec3<Scalar>(box.minImage(vec_to_scalar3(r0 - position + origin)));
+        return Scalar(0.5) * m_k_translational * dot(dr, dr);
+        }
+
+    //! Calculate the energy associated with the deviation of a single particle from its reference
+    //! orientation
+    Scalar calcE_rot(const unsigned int& index, const quat<Scalar>& orientation)
+        {
+        assert(m_symmetry.size());
+        ArrayHandle<unsigned int> h_tags(m_pdata->getTags(),
+                                         access_location::host,
+                                         access_mode::read);
+        quat<Scalar> q0 = m_reference_orientations[h_tags.data[index]];
+        Scalar dqmin = 0.0;
+        for (size_t i = 0; i < m_symmetry.size(); i++)
+            {
+            quat<Scalar> equiv_orientation = orientation * m_symmetry[i];
+            quat<Scalar> dq = q0 - equiv_orientation;
+            dqmin = (i == 0) ? norm2(dq) : fmin(dqmin, norm2(dq));
+            }
+        return Scalar(0.5) * m_k_rotational * dqmin;
+        }
+
+    Scalar calcE_rot(const unsigned int& index, const Shape& shape)
+        {
+        if (!shape.hasOrientation())
+            return Scalar(0.0);
```

Comment:
Add a note in the documentation that the rotational energy is zero for non-orientable particles (such as `hpmc.integrate.Sphere`).

Reply:
On second review, I noticed some issues here. The method only the `calcE_rot(const unsigned int& index, const Shape& shape)` knows whether the shape is orientable. I haven't search the whole codebase, but it appears this method is not even called. `getEnergy`, `energydiff`, and `calculateDeltaE` all call the other signature.

1) Remove `calcE_rot` and see if it compiles, that would verify it is not used.
2) Update the docs to remove the mention of a zero rotational energy. Users are responsible for setting `k_rotational=0` for non-orientable particles.